source/h2_connection.c

```diff
@@ -1186,19 +1197,99 @@ struct aws_h2err s_decoder_on_push_promise(uint32_t stream_id, uint32_t promised
+    uint32_t delta = 0;
+    AWS_ASSERT(target_window >= current_window);
+    delta = (uint32_t)(target_window - current_window);
+    if (delta > AWS_H2_WINDOW_UPDATE_MAX) {
+        /* Cap delta to maximum allowed */
+        delta = AWS_H2_WINDOW_UPDATE_MAX;
+    }
+    return delta;
+}
+
+static int s_connection_send_update_window_if_needed(struct aws_h2_connection *connection, uint64_t window_size) {
+    AWS_PRECONDITION(aws_channel_thread_is_callers_thread(connection->base.channel_slot->channel));
+
+    /**
+     * Only send a WINDOW_UPDATE frame if the connection window is below the threshold
+     * If the pending amount is greater than uin64 max. Probably an unexpected error, ignores it and cap it.
+     */
+    connection->thread_data.pending_window_update_size_thread =
+        aws_add_u64_saturating(connection->thread_data.pending_window_update_size_thread, window_size);
+    if (connection->thread_data.window_size_self > connection->window_size_threshold_to_send_update) {
         CONNECTION_LOGF(
-            ERROR,
+            TRACE,
             connection,
-            "WINDOW_UPDATE frame on connection failed to be sent, error %s",
-            aws_error_name(aws_last_error()));
-        return AWS_OP_ERR;
+            "Ignoring sending connection window update of size%" PRIu64 ". Current size: %" PRIu32
+            ", threshold: %" PRIu32 " pending: %" PRIu64,
+            window_size,
+            connection->thread_data.window_size_self,
+            connection->window_size_threshold_to_send_update,
+            connection->thread_data.pending_window_update_size_thread);
+        return AWS_OP_SUCCESS;
     }
-    aws_h2_connection_enqueue_outgoing_frame(connection, connection_window_update_frame);
-    connection->thread_data.window_size_self += window_size;
+
+    /* Cap the window to AWS_H2_WINDOW_UPDATE_MAX */
+    uint32_t previous_window = connection->thread_data.window_size_self;
+    uint32_t window_delta = aws_h2_calculate_cap_window_update_delta(
+        connection->thread_data.window_size_self, connection->thread_data.pending_window_update_size_thread);
+
+    if (window_delta != connection->thread_data.pending_window_update_size_thread) {
+        CONNECTION_LOGF(
+            DEBUG,
+            (void *)connection,
+            "Capping window update delta from %" PRIu64 " to %" PRIu32,
+            connection->thread_data.pending_window_update_size_thread,
+            window_delta);
+    }
+
+    if (window_delta > 0) {
+        struct aws_h2_frame *connection_window_update_frame =
+            aws_h2_frame_new_window_update(connection->base.alloc, 0, window_delta);
+        if (!connection_window_update_frame) {
+            CONNECTION_LOGF(
+                ERROR,
+                connection,
+                "WINDOW_UPDATE frame on connection failed to be sent, error %s",
+                aws_error_name(aws_last_error()));
+            connection->thread_data.window_size_self = previous_window;
```

Comment:
trivial: maybe this was necessary in some earlier version of your code, but it's not necessary now
```suggestion
```